File: .net & mono/Weed3/DbQuery.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Noear.Weed {
    /**
 * Created by noear on 14-9-5.
 *
 * 查询语句访问类
 *
 */
    public class DbQuery : DbAccess<DbQuery> {

        /*查询语句*/
        public DbQuery(DbContext context) : base(context) {
            
        }
        
        internal DbQuery sql(SQLBuilder sqlBuilder) {
            this.commandText = sqlBuilder.ToString();
            this.paramS.Clear();
            this._weedKey = null;
            foreach (Object p1 in sqlBuilder.paramS) {
                doSet("", p1);
            }

            return this;
        }

        public long insert() {
            return new SQLer().insert(getCommand(), _tran);
        }
        //
        //===========================================
        //
        protected override String getCommandID() {
            return this.commandText;
        }

        protected override Command getCommand() {

            Command cmd = new Command(this.context);

            cmd.key = getCommandID();
            cmd.paramS = this.paramS;

            StringBuilder sb = new StringBuilder(commandText);

            //1.替换schema
            {
                int idx = 0;
                while (true) {
                    idx = sb.IndexOf('$', idx);
                    if (idx > 0) {
                        sb.Replace(idx, idx + 1, context.getSchema());
                        idx++;
                    }
                    else {
                        break;
                    }
                }
            }

            if (this.paramS.Count > 0) {
                
                int paIdx = 0; //参数位置 
                int atIdx = 0; //? in sb位置//到此时已不再有?...(之前已转换为?)
                

                while (true) {
                    atIdx = sb.IndexOf('?', atIdx);
                    if (atIdx > 0) {
                        Variate temp = doGet(paIdx);
                        if (temp == null) {
                            WeedConfig.logException(cmd, new WeedException("缺少参数"));
                        }

                        temp.setName("p" + paIdx);
                        sb.Replace("?", "@" + temp.getName(), atIdx, 1);

                        //atIdx+= strVal.Length;//增加替换后的长度
                        atIdx += temp.getName().Length +1;
                        paIdx++;
                    }
                    else {
                        break;
                    }
                }
            }

            cmd.text = sb.ToString();

            logCommandBuilt(cmd);

            return cmd;
        }
    }
}

File: java/Weed3/src/main/java/org/noear/weed/DbQuery.java
package org.noear.weed;

import java.sql.SQLException;
import java.util.ArrayList;

/**
 * Created by noear on 14-9-5.
 *
 * 查询语句访问类
 *
 * $.tableName  --$ 代表当表db context schema
 * @paramName   --@ 为参数名的开头
 */
public class DbQuery extends DbAccess<DbQuery> {



    public DbQuery(DbContext context)
    {
        super(context);
    }

    public DbQuery sql(SQLBuilder sqlBuilder) {
        this.commandText = sqlBuilder.toString();
        this.paramS.clear();
        this._weedKey = null;
        for (Object p1 : sqlBuilder.paramS) {
            doSet("", p1);
        }

        return this;
    }

    public long insert() throws SQLException
    {
        return new SQLer().insert(getCommand(),_tran);
    }

    @Override
    protected String getCommandID() {
        return this.commandText;
    }

    @Override
    protected Command getCommand() {

        Command cmd = new Command(this.context,_tran);

        cmd.key     = getCommandID();
        cmd.paramS = new ArrayList<>();
        for(Variate v : this.paramS){
            cmd.paramS.add(v.getValue());
        }

        StringBuilder sb = new StringBuilder(commandText);

        //1.替换schema
        int idx=0;
        while (true) {
            idx = sb.indexOf("$",idx);
            if(idx>0) {
                sb.replace(idx, idx + 1, context.getSchema());
                idx++;
            }
            else {
                break;
            }
        }


        cmd.text = sb.toString();


        runCommandBuiltEvent(cmd);

        return cmd;
    }
}
